db/new_seeds/base/add_declarations.rb

```diff
@@ -1,72 +1,209 @@
 # frozen_string_literal: true
 
 require "active_support/testing/time_helpers"
+include ActiveSupport::Testing::TimeHelpers
 
-ActiveRecord::Base.transaction do
-  include ActiveSupport::Testing::TimeHelpers
+# Configuration - adjust batch sizes based on your system capabilities
+LEAD_PROVIDER_BATCH_SIZE = 5
+PARTICIPANT_BATCH_SIZE = 100
+COHORT_BATCH_SIZE = 10
 
-  CpdLeadProvider.joins(:lead_provider).find_each do |cpd_lead_provider|
+# Pre-load all cohorts to avoid repeated queries
+cohorts = Cohort.all.to_a
+
+# Pre-create schedules for all cohorts in bulk
+cohort_ids_without_standard_schedule = cohorts.reject { |c| Finance::Schedule::ECF.exists?(cohort: c, schedule_identifier: "ecf-standard-september") }.map(&:id)
+cohort_ids_without_extended_schedule = cohorts.reject { |c| Finance::Schedule::ECF.exists?(cohort: c, schedule_identifier: "ecf-extended-september") }.map(&:id)
+
+if cohort_ids_without_standard_schedule.any?
+  cohort_ids_without_standard_schedule.each_slice(COHORT_BATCH_SIZE) do |cohort_ids_batch|
+    ActiveRecord::Base.transaction do
+      cohort_ids_batch.each do |cohort_id|
+        cohort = cohorts.find { |c| c.id == cohort_id }
+        FactoryBot.create(:ecf_schedule, cohort:)
+      end
+    end
+  end
+end
+
+if cohort_ids_without_extended_schedule.any?
+  cohort_ids_without_extended_schedule.each_slice(COHORT_BATCH_SIZE) do |cohort_ids_batch|
+    ActiveRecord::Base.transaction do
+      cohort_ids_batch.each do |cohort_id|
+        cohort = cohorts.find { |c| c.id == cohort_id }
+        FactoryBot.create(:ecf_extended_schedule, cohort:)
+      end
+    end
+  end
+end
+
+# Process lead providers in batches
+cpd_lead_providers = CpdLeadProvider.joins(:lead_provider).to_a
+
+cpd_lead_providers.each_slice(LEAD_PROVIDER_BATCH_SIZE) do |cpd_lead_provider_batch|
```

Comment:
do we really need to batch lead providers, there should only be 7?

Reply:
Hi @cwrw I've kept the batch for lead providers loop as the participants loop is inside of it.. so we process participant in batches for a batch of lead providers at the time.